### PR TITLE
flake-modules/new-plugin: init

### PR DESCRIPTION
**Pull request by khaneliman**
Allows quickly creating new plugins to help fulfill plugin creation requests.

### CODE REVIEW

File: flake-modules/dev/new-plugin.py
Comment (MattSturgeon) [inline flake-modules/dev/new-plugin.py:35]: Feels like this may benefit from some unit tests.

One edge case, for example: how does it handle leading `nvim-` prefixes?

Also, don't we normally use kebab-case for plugin names?

File: flake-modules/dev/new-plugin.py
Comment (MattSturgeon) [inline flake-modules/dev/new-plugin.py:58]: As a TODO item, it'd be nice to have an interactive option for any unspecified args

File: flake-modules/dev/new-plugin.py
Comment (MattSturgeon) [inline flake-modules/dev/new-plugin.py:20]: As a TODO item, it'd be nice to have a way to toggle whether a plugin has settings, should call `setup` etc.

Having settings could also mean a `defaults` test case is added.

**Comment by GaetanLepage** [comment]
Good initiative, my rust attempt is likely to stale for a while !

**Comment by khaneliman** [comment]
> Good initiative, my rust attempt is likely to stale for a while !

Oh yeah... sorry forgot about that one

File: flake-modules/dev/new-plugin.py
Comment (khaneliman) [reply flake-modules/dev/new-plugin.py:35]: Yeah, late night brain fart lol.

File: flake-modules/dev/new-plugin.py
Comment (MattSturgeon) [inline flake-modules/dev/new-plugin.py:40]:
```suggestion
        .strip("-")
```
:upside_down_face: https://docs.python.org/3.4/library/stdtypes.html#str.strip

File: flake-modules/dev/new-plugin.py
Comment (MattSturgeon) [inline flake-modules/dev/new-plugin.py:34]: Comment is out of date; it now converts to kebab case.

Wondering if we should use doc-strings too? https://www.programiz.com/python-programming/docstrings

File: flake-modules/dev/new-plugin.py
Comment (MattSturgeon) [inline flake-modules/dev/new-plugin.py:47]: I'm thinking we should extract creating dirs + writing text to a file to a helper fn.

That way the `main()` fn doesn't have to take responsibility for ensuring dirs exist and without having to make these `create_*_file` fns more repetitive.

```py
def write_to_file(file_path, content: str):
    os.makedirs(os.path.dirname(file_path), exist_ok=True)
    with open(file_path, "w") as f:
        f.write(content)
```

File: flake-modules/dev/new-plugin.py
Comment (MattSturgeon) [inline flake-modules/dev/new-plugin.py:36]: `\W` is non-word chars, right? I'm wondering if we'll end up also needing to convert camelCase -> kebab-case? If so we'll need a capture group for `((?<![A-Z])[A-Z])` or something.

**Comment by MattSturgeon** [comment]
> One thing that came up in https://github.com/nix-community/nixvim/pull/2643 made me wonder about how to handle multiple instances of `nvim` in a name. It stripped them all out and made it `remote` but out plugin creation script in nixpkgs generated `remote-nvim-nvim` and the plugin itself refers to itself in lua as `remote-nvim`

I have two thoughts on this.

1. This script doesn't need to be perfect. The contributor should review and tweak the generated file anyway.
2. If we later add an interactive mode, this will be even easier
    (I'm imagining something similar to the GitHub CLIs interactive commands, e.g. `gh repo create` and `gh pr create` - those make use of Charm's bubblegum TUI widgets)
3. We can probably use a smarter regex here; only removing `nvim` if it is at the start/end: `'^nvim|nvim$'`
  Or we could use a `removePrefix('nvim')` and `removeSuffix('nvim')` combo

**Comment by khaneliman** [comment]
Added the TODO items and tweaked replace to only remove leading and trailing `nvim` occurrences so that we can preserve them when it's part of the name.

File: flake-modules/dev/new-plugin.py
Comment (MattSturgeon) [inline flake-modules/dev/new-plugin.py:46]: Can `strip()` accept a regex? If so we could leverage word boundaries to remove some edge-cases. If not, no big deal.

 Maybe also catch the `-` in one pass using `-*`, although that may be less readable than chaining multiple `.strip()` calls...

```suggestion
    return re.sub(r"[\W_]+", "-", input_string).lower().strip(r"\bnvim\b").strip("-")
```

File: flake-modules/dev/new-plugin.py
Comment (GaetanLepage) [inline flake-modules/dev/new-plugin.py:114]:
```suggestion
    create_nix_file(
        file_path=plugin_path,
        template=default_nix_template,
        name=name,
        originalName=args.originalName,
        package=args.package,
    )
```
I tend to write this kind of function calls like this.
This is 100% subjective: feel free to ignore.

File: flake-modules/dev/new-plugin.py
Comment (khaneliman) [reply flake-modules/dev/new-plugin.py]: I went with chaining calls in case stripping `nvim` caused there to be a leftover `-`.

But, looks like `strip` doesn't accept a regex so we have to use `sub`

**Comment by khaneliman** [comment]
Alright, handled cases where `nvim` might be in a word and then formatted the function calls differently.

**Comment by khaneliman** [comment]
@GaetanLepage our checks dont like it though :P

**Comment by GaetanLepage** [comment]
> @GaetanLepage our checks dont like it though :P

Hmm, weird. Have you appended a comma after the last arg ?

**Comment by khaneliman** [comment]
> > @GaetanLepage our checks dont like it though :P
> 
> Hmm, weird. Have you appended a comma after the last arg ?

Ah, I think that was it. I didn't have a trailing comma so the formatter reverted it back.

**Review by MattSturgeon — APPROVED** [review]
All my feedback is addressed. IDK if @GaetanLepage has anything else to add as our resident python guy :grin: 

Note to self for followup work: 

- [ ] Interactive mode
- [ ] Dynamic template sections
- [ ] Document in `CONTRIBUTING`

File: flake-modules/dev/new-plugin.py
Comment (traxys) [inline flake-modules/dev/new-plugin.py:113]: We should maybe get the repo root in order to make those absolute paths? Except if the devshell command automatically cd's at the root, but I'm not sure about that

Comment (khaneliman) [reply flake-modules/dev/new-plugin.py:113]: Oh, are you thinking if someone is `cd`-ing around the repo while in the devshell? Didn't try that..

Comment (khaneliman) [reply flake-modules/dev/new-plugin.py]: Pushed an update to make sure we generate based on the project root.

**Review by traxys — APPROVED** [review]

**Comment by khaneliman** [comment]
@GaetanLepage any other concerns before we merge first implementation?

**Comment by khaneliman** [comment]
@mergify queue

**Pull request merged**
